cmd/influxd/export/export.go

```diff
@@ -0,0 +1,367 @@
+func (cmd *Command) exportResponsePaths(response *snapshotter.Response, tsStart, tsEnd time.Time) error {
+	if err := cmd.exportMetastore(); err != nil {
+		return err
+	}
+
+	// loop through the returned paths and back up each shard
+	for _, path := range response.Paths {
+		rp, id, err := retentionAndShardFromPath(path)
+		if err != nil {
+			return err
+		}
+
+		if err := cmd.exportShard(rp, id, tsStart, tsEnd); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// exportMetastore will backup the metastore on the host to the passed in path. Database and retention policy backups
+// will force a backup of the metastore as well as requesting a specific shard backup from the command line
+func (cmd *Command) exportMetastore() error {
+	metastoreArchivePath, err := cmd.nextPath(filepath.Join(cmd.path, Metafile))
+	if err != nil {
+		return err
+	}
+
+	cmd.StdoutLogger.Printf("backing up metastore to %s", metastoreArchivePath)
+
+	req := &snapshotter.Request{
+		Type: snapshotter.RequestMetastoreBackup,
+	}
+
+	return cmd.downloadAndVerify(req, metastoreArchivePath, func(file string) error {
+		binData, err := ioutil.ReadFile(file)
+		if err != nil {
+			return err
+		}
+
+		magic := binary.BigEndian.Uint64(binData[:8])
+		if magic != snapshotter.BackupMagicHeader {
+			cmd.StderrLogger.Println("Invalid metadata blob, ensure the metadata service is running (default port 8088)")
+			return errors.New("invalid metadata received")
+		}
+
+		return nil
+	})
+}
+
+// nextPath returns the next file to write to.
+func (cmd *Command) nextPath(path string) (string, error) {
+	// Iterate through incremental files until one is available.
+	for i := 0; ; i++ {
+		s := fmt.Sprintf(path+".%02d", i)
+		if _, err := os.Stat(s); os.IsNotExist(err) {
+			return s, nil
+		} else if err != nil {
+			return "", err
+		}
+	}
+}
+
+// downloadAndVerify will download either the metastore or shard to a temp file and then
+// rename it to a good backup file name after complete
+func (cmd *Command) downloadAndVerify(req *snapshotter.Request, path string, validator func(string) error) error {
+	tmppath := path + Suffix
+	if err := cmd.download(req, tmppath); err != nil {
+		return err
+	}
+
+	if validator != nil {
+		if err := validator(tmppath); err != nil {
+			if rmErr := os.Remove(tmppath); rmErr != nil {
+				cmd.StderrLogger.Printf("Error cleaning up temporary file: %v", rmErr)
+			}
+			return err
+		}
+	}
+
+	f, err := os.Stat(tmppath)
+	if err != nil {
+		return err
+	}
+
+	// There was nothing downloaded, don't create an empty backup file.
+	if f.Size() == 0 {
+		return os.Remove(tmppath)
+	}
+
+	// Rename temporary file to final path.
+	if err := os.Rename(tmppath, path); err != nil {
+		return fmt.Errorf("rename: %s", err)
+	}
+
+	return nil
+}
+
+// download downloads a snapshot of either the metastore or a shard from a host to a given path.
+func (cmd *Command) download(req *snapshotter.Request, path string) error {
```

Comment:
A lot of this `download` related code looks copied form the backup command.  Maybe extract that into something re-usable by both?